### ISSUE TITLE
Problem: Number of tracks displayed initially is incorrect

### ISSUE THREAD

**Issue by Robbt** [libretime/libretime#124]
The interface when first started says it is showing 25 tracks but it is limited to only showing 5.

**Comment by Robbt** [comment]
This also happens on the podcast page. I remember this bug on airtime.pro so I will need to fix it as well.

**Comment by Robbt** [comment]
This appears to be a bug in the airtime widget code located in airtime_mvc/public/js/airtime/widgets/table.js
or with how the data is passed to it. The pagination etc also appears to happen in datatables.js
To reproduce on a fresh install, if I upload a number of tracks and then go to Dashboard Tracks it shows only tracks 1-3.

On the podcasts table it limits the podcasts to 25 unless you expand it by selecting a larger range. These maybe separate bugs but I'm currently just trying to untangle the javascript to get a better mental model of whats happening.

**Comment by Robbt** [comment]
I'm going to be looking at the commits around here https://github.com/LibreTime/libretime/commits/276a69aa1bef527260efc170d7caab7d04b4ca13/airtime_mvc/public/js/airtime/widgets/table.js to try to find out how this works and where the bug is. This comment is just for future reference.

**Comment by Robbt** [comment]
@hairmare Do you know offhand where in the code the mechanism for passing data from the database to the DataTables is ? I got stuck trying to find out untangle the mess of various older versions of jquery datatables and plugins, but I know you did some work on this when you build the button to close the "now playing" section of the dashboard, so I thought I'd ask.

**Comment by hairmare** [comment]
I've attempted to find the fix for this on a couple of installs so far. The datatables code is really a tangled mess and nothing so far has helped me figure out where the limit to 5 is coming from.

I was looking at `airtime_mvc/public/js/airtime/library/library.js` at the time, on [line 632](https://github.com/LibreTime/libretime/blob/master/airtime_mvc/public/js/airtime/library/library.js#L632) I found the code that defines the available limits for the table. There is also the [`async: true` case on line 662](https://github.com/LibreTime/libretime/blob/master/airtime_mvc/public/js/airtime/library/library.js#L662) from https://github.com/LibreTime/libretime/commit/0274920d145056c9cb74749783325c75e91fe356. that I did to fix #69. I'm assuming the real action is in `fnServerData ` on [line 708](https://github.com/LibreTime/libretime/blob/master/airtime_mvc/public/js/airtime/library/library.js#L708) but I never figured out where the actual "5" limit was from.

**Comment by Robbt** [comment]
Thanks. at least I know that I'm not the only one scratching my head when it comes to this,  I stared at the code and became bewildered by the nightmare of different versions and libraries and hacks,  indeed it is a tangled scary mess. I'm sure with enough persistence we can figure it out, but not sure if this is the most productive use of my time. I do think getting a better grasp of this will help in any attempt to add new UI functionality in terms of alternative means of organizing the library. 

Something we might want to consider is documenting what different libraries are being used and where. I know that Angular is also used but so is jQuery DataTables and a variety of other libraries are included. 

Maintaining this mess is probably the biggest technical debt we inherited from the project because the only developer documentation that exists is inside of the commit messages or on the now inaccessible sourcefabric wiki. I'll have to look into our documentation set-up and start contributing.

**Comment by hairmare** [comment]
The php library debacle was similar to this before we kicked all the locally committed libs and switched to composer. 

It's just the JS parts are so much worse, I've tried to figure out where to start with a refactor in the direction of using [bower](https://bower.io/) but haven't found a way in so far. 

I'm not that afraid of the angular 1 and jquery parts, those are doable. The hardcore issues seem to be with upgrading datatables and plupload. At least the latter works fine, looks nice and doesn't need any changes at the moment.

**Comment by Robbt** [comment]
Ok, so the data comes from [LibraryController](https://github.com/LibreTime/libretime/blob/7c86991c50effebd7544aef58b38cb84687fdaa4/airtime_mvc/application/controllers/LibraryController.php)
    public function contentsFeedAction()
    {
        $params = $this->getRequest()->getParams();
        # terrible name for the method below. it does not only search files.
        $r = Application_Model_StoredFile::searchLibraryFiles($params);
        $this->view->sEcho = $r["sEcho"];
        $this->view->iTotalDisplayRecords = $r["iTotalDisplayRecords"];
        $this->view->iTotalRecords = $r["iTotalRecords"];
        $this->view->files = SecurityHelper::htmlescape_recursive($r["aaData"]);
    }

whereas     public static function searchLibraryFiles($datatables) is in [StoredFile.php](https://github.com/LibreTime/libretime/blob/fcebafbc6bc8361da2834ddac564dcf01a5ed4bc/airtime_mvc/application/models/StoredFile.php)
but this gets wrapped by [DataTables.php](https://github.com/LibreTime/libretime/blob/276a69aa1bef527260efc170d7caab7d04b4ca13/airtime_mvc/application/models/Datatables.php) which has the iTotalRecords etc.

Not sure how they all play together but these are at least the parts I found based upon digging thanks to you pointing me in the right direction. $params in this case might be built via the DataTables model ?

**Comment by hairmare** [comment]
I'd rather it already came from [`rest/MediaController.php`](https://github.com/LibreTime/libretime/blob/9d5d866db6756e4e7900da014804b472651b13f2/airtime_mvc/application/modules/rest/controllers/MediaController.php), large parts of the pythony stuff and uploading has already been rewritten to use the REST API. That's certainly outside of the scope of this issue and belongs to untangling the frontend mess though.

Yes, `$params` is from the body of the POST request coming from datatables. I'm not sure if it's json or form urlencoded. If it's json I'd assume that the unpacking was either done by the framework or get's done in `Application_Model_StoredFile::searchLibraryFiles()` (which does seem to have side effects 😉 ).

**Comment by hairmare** [comment]
@Robbt FYI, I did lot's of hacking over the weekend and got further than ever before on moving most of `airtime_mc/public/libs` to [bower](https://bower.io/). My work is on https://github.com/radiorabe/libretime/tree/dev/js-refactor and I'll push a refactor of the podcast/library crazyness (with an update to datatables latest version) soonish (maybe by the end of the week).

Also (totally offtopic), I've been preparing the [release notes](https://github.com/LibreTime/libretime/releases) for `3.0.0-alpha.2` and plan on making a PR out of what's part of #173 so we can announce preliminary Ubuntu Xenial support in the release (I did all my dev work this weekend on it and haven't found any distro specific issues since #226).
